File: articles/grit-and-purpose/python/grit_purpose_model.py
"""
Grit and Purpose

Synthetic-data workflow for modeling grit, purpose, support, feedback,
autonomy, opportunity, burnout, and long-term persistence.

This script is for article support and research-method demonstration only.
It should not be used to assess, rank, hire, admit, discipline, or evaluate real people.
"""

from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.api as sm


ROOT = Path(__file__).resolve().parents[1]
DATA_OUT = ROOT / "data" / "processed" / "grit-purpose-modeled-data-python.csv"
TABLE_OUT = ROOT / "outputs" / "tables" / "grit-purpose-model-comparison-python.csv"
PROFILE_OUT = ROOT / "outputs" / "tables" / "grit-purpose-profile-summary-python.csv"
CORR_OUT = ROOT / "outputs" / "tables" / "grit-purpose-correlations-python.csv"

rng = np.random.default_rng(42)
n = 1000

perseverance_effort = rng.normal(0, 1, n)
consistency_interests = rng.normal(0, 1, n)
grit = 0.60 * perseverance_effort + 0.40 * consistency_interests

personal_meaning = rng.normal(0, 1, n)
long_term_direction = rng.normal(0, 1, n)
beyond_self_contribution = rng.normal(0, 1, n)

purpose = (
    0.34 * personal_meaning
    + 0.33 * long_term_direction
    + 0.33 * beyond_self_contribution
)

social_support = rng.normal(0, 1, n)
feedback_quality = rng.normal(0, 1, n)
opportunity_access = rng.normal(0, 1, n)
autonomy_support = rng.normal(0, 1, n)
health_stability = rng.normal(0, 1, n)

burnout = (
    0.18 * grit
    + 0.16 * purpose
    - 0.24 * social_support
    - 0.22 * autonomy_support
    - 0.20 * health_stability
    + rng.normal(0, 1, n)
)

grit_purpose_interaction = grit * purpose

long_term_persistence = (
    0.20 * grit
    + 0.26 * purpose
    + 0.12 * grit_purpose_interaction
    + 0.18 * social_support
    + 0.16 * feedback_quality
    + 0.18 * opportunity_access
    + 0.16 * autonomy_support
    - 0.20 * burnout
    + rng.normal(0, 1, n)
)

df = pd.DataFrame(
    {
        "perseverance_effort": perseverance_effort,
        "consistency_interests": consistency_interests,
        "grit": grit,
        "personal_meaning": personal_meaning,
        "long_term_direction": long_term_direction,
        "beyond_self_contribution": beyond_self_contribution,
        "purpose": purpose,
        "social_support": social_support,
        "feedback_quality": feedback_quality,
        "opportunity_access": opportunity_access,
        "autonomy_support": autonomy_support,
        "health_stability": health_stability,
        "burnout": burnout,
        "grit_purpose_interaction": grit_purpose_interaction,
        "long_term_persistence": long_term_persistence,
    }
)

grit_median = df["grit"].median()
purpose_median = df["purpose"].median()

conditions = [
    (df["grit"] >= grit_median) & (df["purpose"] >= purpose_median),
    (df["grit"] >= grit_median) & (df["purpose"] < purpose_median),
    (df["grit"] < grit_median) & (df["purpose"] >= purpose_median),
    (df["grit"] < grit_median) & (df["purpose"] < purpose_median),
]
choices = [
    "high_grit_high_purpose",
    "high_grit_low_purpose",
    "low_grit_high_purpose",
    "low_grit_low_purpose",
]
df["profile"] = np.select(conditions, choices, default="unclassified")

DATA_OUT.parent.mkdir(parents=True, exist_ok=True)
TABLE_OUT.parent.mkdir(parents=True, exist_ok=True)
PROFILE_OUT.parent.mkdir(parents=True, exist_ok=True)
CORR_OUT.parent.mkdir(parents=True, exist_ok=True)

df.to_csv(DATA_OUT, index=False)

correlations = df[
    [
        "grit",
        "purpose",
        "social_support",
        "feedback_quality",
        "opportunity_access",
        "autonomy_support",
        "health_stability",
        "burnout",
        "long_term_persistence",
    ]
].corr()

correlations.to_csv(CORR_OUT)

model_grit_only = sm.OLS(
    df["long_term_persistence"],
    sm.add_constant(df[["grit"]]),
).fit()

model_grit_purpose = sm.OLS(
    df["long_term_persistence"],
    sm.add_constant(df[["grit", "purpose"]]),
).fit()

model_interaction = sm.OLS(
    df["long_term_persistence"],
    sm.add_constant(df[["grit", "purpose", "grit_purpose_interaction"]]),
).fit()

model_contextual = sm.OLS(
    df["long_term_persistence"],
    sm.add_constant(
        df[
            [
                "grit",
                "purpose",
                "grit_purpose_interaction",
                "social_support",
                "feedback_quality",
                "opportunity_access",
                "autonomy_support",
                "health_stability",
                "burnout",
            ]
        ]
    ),
).fit()

comparison = pd.DataFrame(
    {
        "model": [
            "grit_only",
            "grit_plus_purpose",
            "grit_purpose_interaction",
            "contextual_model",
        ],
        "r_squared": [
            model_grit_only.rsquared,
            model_grit_purpose.rsquared,
            model_interaction.rsquared,
            model_contextual.rsquared,
        ],
        "adjusted_r_squared": [
            model_grit_only.rsquared_adj,
            model_grit_purpose.rsquared_adj,
            model_interaction.rsquared_adj,
            model_contextual.rsquared_adj,
        ],
    }
)

profile_summary = (
    df.groupby("profile", as_index=False)
    .agg(
        n=("profile", "size"),
        mean_long_term_persistence=("long_term_persistence", "mean"),
        mean_grit=("grit", "mean"),
        mean_purpose=("purpose", "mean"),
        mean_social_support=("social_support", "mean"),
        mean_opportunity_access=("opportunity_access", "mean"),
        mean_autonomy_support=("autonomy_support", "mean"),
        mean_health_stability=("health_stability", "mean"),
        mean_burnout=("burnout", "mean"),
    )
)

comparison.to_csv(TABLE_OUT, index=False)
profile_summary.to_csv(PROFILE_OUT, index=False)

print("Saved modeled dataset:", DATA_OUT)
print("Saved correlations:", CORR_OUT)
print("Saved model comparison:", TABLE_OUT)
print("Saved profile summary:", PROFILE_OUT)
print(comparison.round(4))
print(profile_summary.round(4))

File: articles/grit-and-purpose/r/grit-purpose-model.R
# Grit and Purpose
# Synthetic-data workflow for research-method demonstration only.
# Do not use this workflow to evaluate, rank, hire, admit, discipline, or assess real people.

set.seed(42)

n <- 1000

perseverance_effort <- rnorm(n)
consistency_interests <- rnorm(n)
grit <- 0.60 * perseverance_effort + 0.40 * consistency_interests

personal_meaning <- rnorm(n)
long_term_direction <- rnorm(n)
beyond_self_contribution <- rnorm(n)

purpose <- (
  0.34 * personal_meaning +
  0.33 * long_term_direction +
  0.33 * beyond_self_contribution
)

social_support <- rnorm(n)
feedback_quality <- rnorm(n)
opportunity_access <- rnorm(n)
autonomy_support <- rnorm(n)
health_stability <- rnorm(n)

burnout <- (
  0.18 * grit +
  0.16 * purpose -
  0.24 * social_support -
  0.22 * autonomy_support -
  0.20 * health_stability +
  rnorm(n)
)

grit_purpose_interaction <- grit * purpose

long_term_persistence <- (
  0.20 * grit +
  0.26 * purpose +
  0.12 * grit_purpose_interaction +
  0.18 * social_support +
  0.16 * feedback_quality +
  0.18 * opportunity_access +
  0.16 * autonomy_support -
  0.20 * burnout +
  rnorm(n)
)

df <- data.frame(
  perseverance_effort,
  consistency_interests,
  grit,
  personal_meaning,
  long_term_direction,
  beyond_self_contribution,
  purpose,
  social_support,
  feedback_quality,
  opportunity_access,
  autonomy_support,
  health_stability,
  burnout,
  grit_purpose_interaction,
  long_term_persistence
)

grit_median <- median(df$grit)
purpose_median <- median(df$purpose)

df$profile <- ifelse(
  df$grit >= grit_median & df$purpose >= purpose_median,
  "high_grit_high_purpose",
  ifelse(
    df$grit >= grit_median & df$purpose < purpose_median,
    "high_grit_low_purpose",
    ifelse(
      df$grit < grit_median & df$purpose >= purpose_median,
      "low_grit_high_purpose",
      "low_grit_low_purpose"
    )
  )
)

dir.create("data/processed", recursive = TRUE, showWarnings = FALSE)
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)

write.csv(df, "data/processed/grit-purpose-modeled-data-r.csv", row.names = FALSE)

correlations <- cor(df[, c(
  "grit",
  "purpose",
  "social_support",
  "feedback_quality",
  "opportunity_access",
  "autonomy_support",
  "health_stability",
  "burnout",
  "long_term_persistence"
)])

write.csv(correlations, "outputs/tables/grit-purpose-correlations-r.csv")

profile_summary <- aggregate(
  cbind(
    long_term_persistence,
    grit,
    purpose,
    social_support,
    opportunity_access,
    autonomy_support,
    health_stability,
    burnout
  ) ~ profile,
  data = df,
  FUN = mean
)

write.csv(profile_summary, "outputs/tables/grit-purpose-profile-summary-r.csv", row.names = FALSE)

model_grit_only <- lm(long_term_persistence ~ grit, data = df)
model_grit_purpose <- lm(long_term_persistence ~ grit + purpose, data = df)

model_interaction <- lm(
  long_term_persistence ~ grit + purpose + grit_purpose_interaction,
  data = df
)

model_contextual <- lm(
  long_term_persistence ~ grit + purpose + grit_purpose_interaction +
    social_support + feedback_quality + opportunity_access +
    autonomy_support + health_stability + burnout,
  data = df
)

comparison <- data.frame(
  model = c(
    "grit_only",
    "grit_plus_purpose",
    "grit_purpose_interaction",
    "contextual_model"
  ),
  r_squared = c(
    summary(model_grit_only)$r.squared,
    summary(model_grit_purpose)$r.squared,
    summary(model_interaction)$r.squared,
    summary(model_contextual)$r.squared
  ),
  adjusted_r_squared = c(
    summary(model_grit_only)$adj.r.squared,
    summary(model_grit_purpose)$adj.r.squared,
    summary(model_interaction)$adj.r.squared,
    summary(model_contextual)$adj.r.squared
  )
)

write.csv(comparison, "outputs/tables/grit-purpose-model-comparison-r.csv", row.names = FALSE)
write.csv(
  summary(model_contextual)$coefficients,
  "outputs/tables/grit-purpose-contextual-coefficients-r.csv"
)

print(round(correlations, 3))
print(round(profile_summary, 3))
print(round(comparison, 4))
print(round(summary(model_contextual)$coefficients, 4))
